lib/Dialect/FIRRTL/FFI/FIRRTLFFIContext.cpp

```diff
@@ -0,0 +1,1683 @@
+  auto *mlirCtx = this->mlirCtx.get();
+
+  FIRRTLType firType;
+
+  switch (type.kind) {
+  case FIRRTL_TYPE_KIND_UINT:
+    firType = UIntType::get(mlirCtx, type.u.uint.width);
+    break;
+  case FIRRTL_TYPE_KIND_SINT:
+    firType = SIntType::get(mlirCtx, type.u.sint.width);
+    break;
+  case FIRRTL_TYPE_KIND_CLOCK:
+    firType = ClockType::get(mlirCtx);
+    break;
+  case FIRRTL_TYPE_KIND_RESET:
+    firType = ResetType::get(mlirCtx);
+    break;
+  case FIRRTL_TYPE_KIND_ASYNC_RESET:
+    firType = AsyncResetType::get(mlirCtx);
+    break;
+  case FIRRTL_TYPE_KIND_ANALOG:
+    firType = AnalogType::get(mlirCtx, type.u.analog.width);
+    break;
+  case FIRRTL_TYPE_KIND_VECTOR: {
+    auto elementType = ffiTypeToFirType(*type.u.vector.type);
+    if (!elementType.has_value()) {
+      return std::nullopt;
+    }
+    auto baseType = elementType->dyn_cast<FIRRTLBaseType>();
+    if (!baseType) {
+      emitError("element must be base type");
+      return std::nullopt;
+    }
+
+    firType = FVectorType::get(baseType, type.u.vector.count);
+    break;
+  }
+  case FIRRTL_TYPE_KIND_BUNDLE: {
+    SmallVector<BundleType::BundleElement, 4> fields;
+    fields.reserve(type.u.bundle.count);
+
+    for (size_t i = 0; i < type.u.bundle.count; i++) {
+      const auto &field = type.u.bundle.fields[i];
+
+      auto fieldType = ffiTypeToFirType(*field.type);
+      if (!fieldType.has_value()) {
+        return std::nullopt;
+      }
+      auto baseType = fieldType->dyn_cast<FIRRTLBaseType>();
+      if (!baseType) {
+        emitError("field must be base type");
+        return std::nullopt;
+      }
+
+      fields.emplace_back(stringRefToAttr(unwrap(field.name)), field.flip,
+                          baseType);
+    }
+    firType = BundleType::get(mlirCtx, fields);
+    break;
+  }
+  default: // NOLINT(clang-diagnostic-covered-switch-default)
+    emitError("unknown type kind");
+    return std::nullopt;
+  }
+
+  return firType;
+}
+
+bool FFIContext::checkFinal() const {
+  if (!whenStack.empty()) {
+    emitError("expected " + std::to_string(whenStack.size()) + " WhenEnd");
+    return false;
+  }
+  return true;
+}
+
+std::optional<std::pair<firrtl::FModuleOp &, Block *>>
+FFIContext::currentModule() {
+  auto *moduleOpPtr =
+      std::get_if<details::RequireAssigned<firrtl::FModuleOp>>(&this->moduleOp);
+  if (moduleOpPtr == nullptr) {
+    emitError("declaration or statement cannot be placed under an `extmodule`");
+    return std::nullopt;
+  }
+  auto &moduleOp = *moduleOpPtr;
+  RA_EXPECT(auto &lastModuleOp, moduleOp, std::nullopt);
+
+  Block *blockToInsertInto;
+  if (!whenStack.empty()) {
+    blockToInsertInto = &whenStack.top().currentBlock();
+  } else {
+    blockToInsertInto = lastModuleOp.getBodyBlock();
+  }
+
+  return std::make_pair(std::ref(lastModuleOp), blockToInsertInto);
+}
+
+// NOLINTNEXTLINE(misc-no-recursion)
+std::optional<mlir::Value> FFIContext::resolveRef(BodyOpBuilder &bodyOpBuilder,
```

Comment:
The CAPI shouldn't be parsing strings.  There are operations which a client can construct to index the strings based on it's parsing.